File: ex400.java
class Solution {
    public int findNthDigit(int n) {
        if (n == Math.pow(2, 31) - 1)    return 2;
        
        int i = 1;
        int numOfDigit = 1;
        int backup = 0;
        int enlongate = 9, num = 9;
        
        while (true) {
            backup++;
            if (i >= n)    break;
            
            if (backup >= enlongate) {
                numOfDigit++;
                num *= 10;
                enlongate += num;
            }
            i += numOfDigit;
        }
        i -= numOfDigit;
        
        return String.valueOf(backup).charAt(n - i - 1) - '0';
    }
}

File: code/ex400.cpp
class Solution {
public:
    int findNthDigit(int n) {
        if (n < 10)    return n;
        
        // Get the digits of the number where nth digit is in
        n -= 9;
        int digits = 2;
        while (n - (pow(10, digits) - pow(10, digits - 1)) * digits > 0) {
            n -= (pow(10, digits) - pow(10, digits - 1)) * digits;
            digits++;
        }
        
        // Get the number this digit is in
        n--;
        long long num = n / digits + pow(10, digits - 1);
        int d = n % digits;    // The digit in the num

        return to_string(num)[d] - '0';
    }
};